mixer/adapter/kubernetes2/kubernetes.go

```diff
@@ -0,0 +1,401 @@
+	}
+	for i := len(nameParts); i < len(domParts); i++ {
+		s = s + "." + domParts[i]
+	}
+	return s, nil
+}
+
+func keyFromUID(uid string) string {
+	if ip := net.ParseIP(uid); ip != nil {
+		return uid
+	}
+	fullname := strings.TrimPrefix(uid, kubePrefix)
+	if strings.Contains(fullname, ".") {
+		parts := strings.Split(fullname, ".")
+		if len(parts) == 2 {
+			return key(parts[1], parts[0])
+		}
+	}
+	return fullname
+}
+
+func fillOriginAttrs(p *v1.Pod, o *ktmpl.Output, params *config.Params) {
+	if len(p.Labels) > 0 {
+		o.OriginLabels = p.Labels
+	}
+	if len(p.Name) > 0 {
+		o.OriginPodName = p.Name
+	}
+	if len(p.Namespace) > 0 {
+		o.OriginNamespace = p.Namespace
+	}
+	if len(p.Spec.ServiceAccountName) > 0 {
+		o.OriginServiceAccountName = p.Spec.ServiceAccountName
+	}
+	if len(p.Status.PodIP) > 0 {
+		o.OriginPodIp = net.ParseIP(p.Status.PodIP)
+	}
+	if len(p.Status.HostIP) > 0 {
+		o.OriginHostIp = net.ParseIP(p.Status.HostIP)
+	}
+	if app, found := p.Labels[params.PodLabelForService]; found {
+		n, err := canonicalName(app, p.Namespace, params.ClusterDomainName)
+		if err == nil {
+			o.OriginService = n
+		}
+	} else if app, found := p.Labels[params.PodLabelForIstioComponentService]; found {
+		n, err := canonicalName(app, p.Namespace, params.ClusterDomainName)
+		if err == nil {
+			o.OriginService = n
+		}
+	}
+}
+
+func fillDestinationAttrs(p *v1.Pod, o *ktmpl.Output, params *config.Params) {
+	if len(p.Labels) > 0 {
+		o.DestinationLabels = p.Labels
+	}
+	if len(p.Name) > 0 {
+		o.DestinationPodName = p.Name
+	}
+	if len(p.Namespace) > 0 {
+		o.DestinationNamespace = p.Namespace
+	}
+	if len(p.Spec.ServiceAccountName) > 0 {
+		o.DestinationServiceAccountName = p.Spec.ServiceAccountName
+	}
+	if len(p.Status.PodIP) > 0 {
+		o.DestinationPodIp = net.ParseIP(p.Status.PodIP)
+	}
+	if len(p.Status.HostIP) > 0 {
+		o.DestinationHostIp = net.ParseIP(p.Status.HostIP)
+	}
+	if app, found := p.Labels[params.PodLabelForService]; found {
+		n, err := canonicalName(app, p.Namespace, params.ClusterDomainName)
+		if err == nil {
+			o.DestinationService = n
+		}
+	} else if app, found := p.Labels[params.PodLabelForIstioComponentService]; found {
+		n, err := canonicalName(app, p.Namespace, params.ClusterDomainName)
+		if err == nil {
+			o.DestinationService = n
+		}
+	}
+}
+
+func fillSourceAttrs(p *v1.Pod, o *ktmpl.Output, params *config.Params) {
+	if len(p.Labels) > 0 {
+		o.SourceLabels = p.Labels
+	}
+	if len(p.Name) > 0 {
+		o.SourcePodName = p.Name
+	}
+	if len(p.Namespace) > 0 {
+		o.SourceNamespace = p.Namespace
+	}
+	if len(p.Spec.ServiceAccountName) > 0 {
+		o.SourceServiceAccountName = p.Spec.ServiceAccountName
+	}
+	if len(p.Status.PodIP) > 0 {
+		o.SourcePodIp = net.ParseIP(p.Status.PodIP)
+	}
+	if len(p.Status.HostIP) > 0 {
+		o.SourceHostIp = net.ParseIP(p.Status.HostIP)
+	}
+	if app, found := p.Labels[params.PodLabelForService]; found {
+		n, err := canonicalName(app, p.Namespace, params.ClusterDomainName)
+		if err == nil {
```

Comment:
What about when err != nil? We're not handling the failure case. Same for the code below. Should we at least log something?